### ISSUE TITLE
[API Proposal]: Create a ToListOrEmpty<T> on System.Collections.Generic

### ISSUE THREAD

**Issue by ElectNewt** [dotnet/runtime#105660]
### Background and motivation

`ToListOrEmpty<T>` is an extension method that i saw repeatly during most of my career.  It has been used mainly when getting information from 3rd party applications where I expect a response, but for whatever reason it might not be there (normally it will be null) and I dont want the process to break in that case.

This issue is easily solvale with the following code: 

```
IEnumerable<Foo>? foo = _provider.GetFoo();
List<Foo> foolist = foo?.ToList() ?? new List<Foo>(); //or foo?.ToList() ?? []
```

Which is not too bad, but as I had to write/review this so many times, I rather to have a method for it. 

I searched for similar issues but i could not find any, and I thought it will be worth the shot to study if it is worth to add it, specially for those migrating from legacy systems (mainly .NET Framework) to .NET 8+ as I saw this situation repeated on those kind of codebases.  At the same time its an extra abstraction to maintain which makes me wonder if this is really worth the effort.

### API Proposal

```csharp
namespace System.Collections.Generic;

namespace System.Linq
{
    public static partial class Enumerable
    {


     public static List<TSource> ToListOrEmpty<TSource>(this IEnumerable<TSource>? source){}

     }
}
```

### API Usage

```csharp
IEnumerable<Foo> foo = _provider.GetFoo();

List<Foo> foolist = foo.ToListOrEmpty();
```

**Comment by colejohnson66** [comment]
I disagree with this. Many times, people are using null to indicate a lack of results. Instead, what they should be returning is `Enumerable.Empty<T>()`, or `Array.Empty<T>()`.

The proposed API just feeds bad practices.

**Comment by huoyaoyuan** [comment]
> Many times, people are using null to indicate a lack of results. Instead, what they should be returning is `Enumerable.Empty<T>()`, or `Array.Empty<T>()`.

The source can be out of control from the usage site.

But we don't provide `IEnumerable<T>?` handling in LINQ. `ToList` is just one single case among LINQ. We won't add a single case for this.

**Comment by ElectNewt** [comment]
> Many times, people are using null to indicate a lack of results. Instead, what they should be returning is `Enumerable.Empty<T>()`, or `Array.Empty<T>()`.

yes, I agree that sometimes null means a lack of results and for that scenario, this API is not needed. but the scenarios I've been facing the last few years with people migrating/dealing with legacy systems is that you have to deal with this scenario.

about the `Enumerable.Empty<T>()`, or `Array.Empty<T>()` when i mentioned 3rd party providers I meant services (usually other companies) that you have no access at all.

**Comment by wvpm** [comment]
`foo?.ToList() ?? []` is immediately clear to me.
`foo.ToListOrEmpty()` is not. It raises the question "Why not `.ToList()`?" (because it might be null).

Besides that, I also wonder why you want List. Do you intend to change the length? If not, `.ToArray()` is more appropriate.

**Comment by julealgon** [comment]
I've worked on a project where there were "Safe" variations of tons of LINQ methods, so that they could be called "safely" on `null` values. These ended up infesting the entire codebase with thousands of usages each.

This proposal basically reminds me of that nightmare. It just promotes a terrible practice. I'd hate to see such methods in the BCL for things that lead people into the pit of failure by default.

**Comment by rjgotten** [comment]
> Besides that, I also wonder why you want List. Do you intend to change the length? If not, `.ToArray()` is more appropriate.

Actually the most appropriate is:

```cs
return [.. foo ?? []];
```
Which will allow the compiler to select an optimized implementation under the hood.
Currently `ReadOnlyArray<>` if your method returns either `IEnumerable<>` or `IReadOnlyList<>`, iirc.

**Comment by DrkWzrd** [comment]
This should come with (or refactored to) another convenience method

```c#
public static IEnumerable<T> NotNull<T>(this IEnumerable<T>? possiblyNull)
{
    return possiblyNull ?? [];
}
```

Even more, with this previous method I think is more readable

```c#
IEnumerable<> data = GetPossiblyNullData();
var list = data.NotNull().ToList();
```

**Comment by wvpm** [comment]
> This should come with (or refactored to) another convenience method
> 
> ```cs
> public static IEnumerable<T> NotNull<T>(this IEnumerable<T>? possiblyNull)
> {
>     return possiblyNull ?? [];
> }
> ```
> 
> I think this should be the only method managing null `IEnumerable` in linq. And it is the simplest way to protect from `NullReferenceException` in linq.
> 
> Even more, as @wvpm says, with this previous method I think is more readable
> 
> ```cs
> IEnumerable<> data = GetPossiblyNullData();
> var list = data.NotNull().ToList();
> 
> //event you can extend it to linq queries
> var elements = from value in /*not null*/ data select...
> ```

Since you mention linq queries, I'll focus on that type of usage of Linq.
Linq queries, in my opinion, should be used for querying.
That is:

- joining 'tables' (via `Select`)
- declaring which columns you want (via `Select`)
- expanding collections (via `SelectMany`)
- aggregation (via `Sum`, `Count`, `Average` & `Aggregate`)
- filtering (`Where`)
- paging (`Skip` & `Take`)
- ordering/sorting (`OrderBy`)

The `ToList` & `ToArray` are not part of the querying but used to execute the query and materialise the `IEnumerable`. Including a `not null` in the linq query therefore feels out of place.

As others have pointed out, a query returning `null` is strange behaviour. It does not make sense in the context of querying a database. `IEnumerable<T>?` tends to appear when querying a collection that is nullable.

If the source is not null, a linq query doesn't produce `null`. Therefore handling the `null` case does not belong to the responsibility of linq queries. Handling the `null` case should be done outside of linq code.

Suppose you want to do the following:
```csharp
var obj = new {
	Collection = (IEnumerable<string>?) null
};
string[] strings = obj.Collection.Where(x => x.Length > 0).ToArray();
```

Prefer:
`string[] strings = obj.Collection is {} collection ? collection.Where(x => x.Length > 0).ToArray() : null`
Shorthand:
`string[] strings = obj?.Collection.Where(x => x.Length > 0).ToArray() ?? []`

Over
`string[] strings = obj.Collection.Where(x => x.Length > 0).NotNull().ToArray()`
or
`string[] strings = obj.Collection.Where(x => x.Length > 0).ToArrayOrEmpty()`

Methods such as `Where` specifically throw `ArgumentNullException`, so Linq has no reason to handle the `null` case.

**Comment by DrkWzrd** [comment]
> Over `string[] strings = obj.Collection.Where(x => x.Length > 0).NotNull().ToArray()` or `string[] strings = obj.Collection.Where(x => x.Length > 0).ToArrayOrEmpty()`
> 
> Methods such as `Where` specifically throw `ArgumentNullException`, so Linq has no reason to handle the `null` case.

Here I think this should come like this
```c#
string[] strings = obj.Collection.AsNotNull().Where(x=> x.Length> 0).ToArray(); //this way you don't care if collection is null, you will get the result or an empty (but enumerable) array. The same result as you are proposing.
```

You are doing an exercise of responsability saying "I know this could be null, and I'm handling it". And you are avoiding throwing an exception, which we all know is preferable.

In the end, these are very subtle (and therefore irrelevant) differences.

**Comment by wvpm** [comment]
@DrkWzrd
Moving the `AsNotNull()` before the query makes more sense and looks better than the previous proposed API changes.
Personally I'd still prefer `?? []` in scenarios like that.
You could also use in this way: `string[] strings = (obj.Collection ?? []).Where(x=> x.Length> 0).ToArray(); `

**Comment by julealgon** [comment]
@DrkWzrd 
> You are doing an exercise of responsability saying "I know this could be null, and I'm handling it". And you are avoiding throwing an exception, which we all know is preferable.

In the vast majority of cases from my experience, a `null` collection is just a bug/bad practice that needs fixing. So, I disagree: handling the `null` value in those cases instead of letting it throw just masks and postpones an issue that needed immediate attention instead.
